// kle-serial-alt/dist/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  Key: () => Key,
  Keyboard: () => Keyboard,
  KeyboardMetadata: () => KeyboardMetadata,
  KleDeserializeError: () => KleDeserializeError,
  Serial: () => Serial,
  deserialize: () => deserialize,
  parse: () => parse
});
module.exports = __toCommonJS(index_exports);
var import_json5 = __toESM(require("json5"), 1);
var Key = class {
  color = "#cccccc";
  labels = [];
  textColor = [];
  textSize = [];
  default = { textColor: "#000000", textSize: 3 };
  x = 0;
  y = 0;
  width = 1;
  height = 1;
  x2 = 0;
  y2 = 0;
  width2 = 1;
  height2 = 1;
  rotation_x = 0;
  rotation_y = 0;
  rotation_angle = 0;
  decal = false;
  ghost = false;
  stepped = false;
  nub = false;
  profile = "";
  sm = "";
  sb = "";
  st = "";
};
var KeyboardMetadata = class {
  author = "";
  backcolor = "#eeeeee";
  background = null;
  name = "";
  notes = "";
  radii = "";
  switchBrand = "";
  switchMount = "";
  switchType = "";
  pcb = false;
  plate = false;
  css = "";
};
var Keyboard = class {
  meta = new KeyboardMetadata();
  keys = [];
};
var LABEL_MAP = [
  [0, 6, 2, 8, 9, 11, 3, 5, 1, 4, 7, 10],
  [1, 7, -1, -1, 9, 11, 4, -1, -1, -1, -1, 10],
  [3, -1, 5, -1, 9, 11, -1, -1, 4, -1, -1, 10],
  [4, -1, -1, -1, 9, 11, -1, -1, -1, -1, -1, 10],
  [0, 6, 2, 8, 10, -1, 3, 5, 1, 4, 7, -1],
  [1, 7, -1, -1, 10, -1, 4, -1, -1, -1, -1, -1],
  [3, -1, 5, -1, 10, -1, -1, -1, 4, -1, -1, -1],
  [4, -1, -1, -1, 10, -1, -1, -1, -1, -1, -1, -1]
];
var reorderLabelsIn = (labels, align, def = null) => {
  const ret = [];
  for (let i = 0; i < labels.length; i += 1) {
    const v = labels[i];
    if (v && v !== def) {
      const idx = LABEL_MAP[align]?.[i];
      if (idx !== void 0 && idx >= 0) ret[idx] = v;
    }
  }
  return ret;
};
var copyKey = (src) => {
  const dst = new Key();
  dst.color = src.color;
  dst.labels = src.labels.slice();
  dst.textColor = src.textColor.slice();
  dst.textSize = src.textSize.slice();
  dst.default = { ...src.default };
  dst.x = src.x;
  dst.y = src.y;
  dst.width = src.width;
  dst.height = src.height;
  dst.x2 = src.x2;
  dst.y2 = src.y2;
  dst.width2 = src.width2;
  dst.height2 = src.height2;
  dst.rotation_x = src.rotation_x;
  dst.rotation_y = src.rotation_y;
  dst.rotation_angle = src.rotation_angle;
  dst.decal = src.decal;
  dst.ghost = src.ghost;
  dst.stepped = src.stepped;
  dst.nub = src.nub;
  dst.profile = src.profile;
  dst.sm = src.sm;
  dst.sb = src.sb;
  dst.st = src.st;
  return dst;
};
var KleDeserializeError = class extends Error {
  constructor(message, data) {
    super(message);
    this.data = data;
    this.name = "KleDeserializeError";
  }
  data;
};
var deserializeError = (msg, data) => {
  const detail = data !== void 0 ? `${msg}:
  ${import_json5.default.stringify(data)}` : msg;
  throw new KleDeserializeError(detail, data);
};
var deserialize = (rows) => {
  if (!Array.isArray(rows)) deserializeError("expected an array of objects");
  let current = new Key();
  const kbd = new Keyboard();
  let align = 4;
  const rowsArr = rows;
  for (let r = 0; r < rowsArr.length; r += 1) {
    const row = rowsArr[r];
    if (Array.isArray(row)) {
      for (let k = 0; k < row.length; k += 1) {
        const item = row[k];
        if (typeof item === "string") {
          const newKey = copyKey(current);
          newKey.width2 = current.width2 === 0 ? current.width : current.width2;
          newKey.height2 = current.height2 === 0 ? current.height : current.height2;
          newKey.labels = reorderLabelsIn(item.split("\n"), align);
          newKey.textSize = reorderLabelsIn(newKey.textSize, align);
          for (let i = 0; i < 12; i += 1) {
            if (!newKey.labels[i]) {
              delete newKey.textSize[i];
              delete newKey.textColor[i];
            }
            if (newKey.textSize[i] === newKey.default.textSize) delete newKey.textSize[i];
            if (newKey.textColor[i] === newKey.default.textColor) delete newKey.textColor[i];
          }
          kbd.keys.push(newKey);
          current.x += current.width;
          current.width = current.height = 1;
          current.x2 = current.y2 = 0;
          current.width2 = current.height2 = 0;
          current.nub = current.stepped = current.decal = false;
        } else if (item && typeof item === "object") {
          const p = item;
          if (k !== 0 && (p.r !== void 0 || p.rx !== void 0 || p.ry !== void 0)) {
            deserializeError("rotation can only be specified on the first key in a row", p);
          }
          if (p.r !== void 0) current.rotation_angle = p.r;
          if (p.rx !== void 0) {
            current.rotation_x = p.rx;
            current.x = p.rx;
            current.y = current.rotation_y;
          }
          if (p.ry !== void 0) {
            current.rotation_y = p.ry;
            current.y = p.ry;
            current.x = current.rotation_x;
          }
          if (p.a !== void 0) align = p.a;
          if (p.f) {
            current.default.textSize = p.f;
            current.textSize = [];
          }
          if (p.f2 !== void 0) {
            for (let i = 1; i < 12; i += 1) current.textSize[i] = p.f2;
          }
          if (p.fa !== void 0) current.textSize = p.fa.slice();
          if (p.p !== void 0) current.profile = p.p;
          if (p.c !== void 0) current.color = p.c;
          if (p.t !== void 0) {
            const split = p.t.split("\n");
            current.default.textColor = split[0];
            current.textColor = reorderLabelsIn(split, align, current.default.textColor);
          }
          if (p.x !== void 0) current.x += p.x;
          if (p.y !== void 0) current.y += p.y;
          if (p.w !== void 0) {
            current.width = p.w;
            current.width2 = p.w;
          }
          if (p.h !== void 0) {
            current.height = p.h;
            current.height2 = p.h;
          }
          if (p.x2 !== void 0) current.x2 = p.x2;
          if (p.y2 !== void 0) current.y2 = p.y2;
          if (p.w2 !== void 0) current.width2 = p.w2;
          if (p.h2 !== void 0) current.height2 = p.h2;
          if (p.n !== void 0) current.nub = !!p.n;
          if (p.l !== void 0) current.stepped = !!p.l;
          if (p.d !== void 0) current.decal = !!p.d;
          if (p.g !== void 0) current.ghost = !!p.g;
          if (p.sm !== void 0) current.sm = p.sm;
          if (p.sb !== void 0) current.sb = p.sb;
          if (p.st !== void 0) current.st = p.st;
        }
      }
      current.y += 1;
      current.x = current.rotation_x;
    } else if (row && typeof row === "object") {
      if (r !== 0) {
        deserializeError("keyboard metadata must be the first element", row);
      }
      const meta = row;
      const out = kbd.meta;
      for (const prop of Object.keys(out)) {
        if (meta[prop] !== void 0 && meta[prop] !== null) {
          out[prop] = meta[prop];
        }
      }
    }
  }
  return kbd;
};
var parse = (json) => deserialize(import_json5.default.parse(json));
var Serial = { deserialize, parse };
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Key,
  Keyboard,
  KeyboardMetadata,
  KleDeserializeError,
  Serial,
  deserialize,
  parse
});
